jni/src/common/Foreign/JNI/Unsafe.hs

```diff
@@ -453,26 +413,35 @@ useAsCStrings strs m =
 -- practice: use 'withJVM' instead. Only useful for GHCi.
 newJVM :: [ByteString] -> IO JVM
 newJVM options = JVM_ <$> do
-    useAsCStrings options $ \cstrs -> do
-      withArray cstrs $ \(coptions :: Ptr (Ptr CChar)) -> do
-        let n = fromIntegral (length cstrs) :: C.CInt
-        checkBoundness
-        jvm <- [C.block| JavaVM * {
-          JavaVM *jvm;
-          JavaVMInitArgs vm_args;
-          JavaVMOption *options = malloc(sizeof(JavaVMOption) * $(int n));
-          for(int i = 0; i < $(int n); i++)
-                  options[i].optionString = $(char **coptions)[i];
-          vm_args.version = JNI_VERSION_1_6;
-          vm_args.nOptions = $(int n);
-          vm_args.options = options;
-          vm_args.ignoreUnrecognized = 0;
-          JNI_CreateJavaVM(&jvm, (void**)&jniEnv, &vm_args);
-          free(options);
-          return jvm; } |]
-        cleaner <- createGlobalRefCleaner
-        writeIORef globalRefCleaner cleaner
-        return jvm
+    checkBoundness
+    startJVM options <* startFinalizerThread
+  where
+    startFinalizerThread =
+      BackgroundWorker.create
+          (exitOnError . runInBoundThread . runInAttachedThread)
+      >>= writeIORef finalizerThread
+
+    exitOnError = handle $ \(e :: SomeException) -> do
+      System.Exit.die $ "Haskell jni package: error in finalizer thread: " ++ show e
```

Comment:
Do we really want to die here, instead of letting the user the opportunity to catch something ?
I could see users wanting to `try withJVM`, at least for debugging purposes.

Reply:
It would be fine to have `withJVM` throw an exception, but I think it would be difficult to make it happen. Would you use asynchronous exceptions to try to stop the main thread? I would prefer to not face the many opportunities for mistakes that asynchronous exceptions can introduce.